packages/sanity/src/core/form/field/FieldActions.tsx

```diff
@@ -0,0 +1,106 @@
+import {Path, SchemaType} from '@sanity/types'
+import React, {memo, useCallback, useEffect, useRef, useState} from 'react'
+import {DocumentFieldAction, DocumentFieldActionNode} from '../../config'
+import {useUnique} from '../../util'
+
+/** @internal */
+export interface FieldActionsProps {
+  actions: DocumentFieldAction[]
+  documentId: string
+  documentType: string
+  onActions: (actions: DocumentFieldActionNode[]) => void
+  path: Path
+  schemaType: SchemaType
+}
+
+/** @internal */
+export const FieldActions = memo(function FieldActions(props: FieldActionsProps) {
+  const {actions, documentId, documentType, onActions, path, schemaType} = props
+
+  const len = actions.length
+  const lenRef = useRef(len)
+
+  const [fieldActions, setFieldActions] = useState<DocumentFieldActionNode[]>(() =>
+    Array.from(new Array(len))
+  )
+
+  const fieldActionsRef = useRef(fieldActions)
+
+  useEffect(() => {
+    fieldActionsRef.current = fieldActions
+  }, [fieldActions])
+
+  useEffect(() => {
+    if (lenRef.current !== len) {
+      const newFieldActions = Array.from(new Array(len))
+
+      for (let i = 0; i < len; i++) {
+        newFieldActions[i] = fieldActionsRef.current[i]
+      }
+
+      lenRef.current = len
+
+      setFieldActions(newFieldActions)
+      fieldActionsRef.current = newFieldActions
+    }
+  }, [len])
+
+  const setFieldAction = useCallback((index: number, node: DocumentFieldActionNode) => {
+    setFieldActions((prev) => {
+      const next = [...prev]
+      next[index] = node
+      return next
+    })
+  }, [])
+
+  useEffect(() => {
+    onActions(fieldActions.filter(Boolean))
+  }, [fieldActions, onActions])
+
+  return (
+    <>
+      {actions.map((a, aIdx) => (
+        <FieldAction
+          action={a}
+          index={aIdx}
+          // eslint-disable-next-line react/no-array-index-key
+          key={aIdx}
+          documentId={documentId}
+          documentType={documentType}
+          path={path}
+          schemaType={schemaType}
+          setFieldAction={setFieldAction}
+        />
+      ))}
+    </>
+  )
+})
+
+interface FieldActionProps {
+  action: DocumentFieldAction
+  documentId: string
+  documentType: string
+  index: number
+  path: Path
+  schemaType: SchemaType
+  setFieldAction: (index: number, node: DocumentFieldActionNode) => void
+}
+
+const FieldAction = memo(function FieldAction(props: FieldActionProps) {
+  const {action, documentId, documentType, index, path, schemaType, setFieldAction} = props
+
+  const node = useUnique(
+    action.useAction({
+      documentId,
+      documentType,
+      path,
+      schemaType,
+    })
+  )
+
+  useEffect(() => {
+    setFieldAction(index, node)
```

Comment:
I wasn't able to reason about the document actions components for resolving hooks, and thought it could be solved in a more straight-forward manner.

This implementation uses a deep compare to prevent render loops. However, it's required to wrap the `onAction` callback in `useCallback`.

Reply:
> This implementation uses a deep compare to prevent render loops

My question was whether an action that returns a value that changes at at every call (e.g a timestamp) can lead to infinite render loops. Maybe an issue more in theory than in practice, but something to keep in mind at least.

Since we support returning jsx from the action (only `disabled.reason` currently, but likely more in the future) we'll also risk doing deep comparison of (potentially) large data structures.

Reply:
True, so this may be something to keep an eye on. Not sure how to solve it, though.

Reply:
(It's `@internal` APIs.)